examples/manifold_gradients.dx

```diff
@@ -0,0 +1,342 @@
+' ## Gradients on manifolds
+An example of how to generalize the Dex `linearize` and `grad` functions to take
+derivatives and gradients of functions on manifolds.
+
+' ### Tangent space
+The difference between differentiating on $\mathbb{R}^n$ and on general
+manifolds is that in the general case, the manifold tangent space type may be
+different to the manifold representation type. The tangent space at a single
+point will always be isomorphic to $\mathbb{R}^n$ for some $n$, but here we
+allow any vector space type `b` for flexibility.
+
+' The proper definition of a tangent space is quite abstract, so no further details will be provided here. Wikipedia has quite a good
+[article](https://en.wikipedia.org/wiki/Tangent_space) for more information.
+
+' Practically, the two components required for defining the differentiable
+structure on the manifold are the exponential map (a mapping from the tangent
+space at p to the manifold, such that zero is mapped to $p$, and straight lines
+are mapped to geodesics), and its inverse. Note that "exponential map" is a
+somewhat overloaded term, here we are talking about it in a Riemannian geometry
+sense.
+
+' In the `TangentSpace` interface, `a` is the type which can represent points
+on the manifold. `b` is the type which can represent points in a tangent space
+to the manifold.  Ideally `b` would be an associated type of `a` so that the
+compiler is able to infer the tangent space type from the manifold type, but
+this feature is not yet available in Dex.
+
+interface [VSpace b] TangentSpace a b
```

Comment:
How about renaming `a` to `manifold` and `b` to `tangent`? Type variables are often single-letters when there is no good generic name (e.g. when you define a `map`, lists can contain anything), but here we do have some very concrete expectations about them. The type of `tangentExponentialMap` would be much easier to read after the renaming.